src/usr.sbin/wsfontload/wsfontload.8
.\"	$MirOS: src/usr.sbin/wsfontload/wsfontload.8,v 1.3 2005/12/04 15:02:33 tg Exp $ */
.\"	$OpenBSD: wsfontload.8,v 1.15 2004/08/03 09:45:29 jmc Exp $
.\"	$NetBSD: wsfontload.8,v 1.5 1999/04/06 04:54:22 cgd Exp $
.\"
.\" Copyright (c) 1999, 2001
.\" 	Matthias Drochner.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd $Mdocdate$
.Dt WSFONTLOAD 8
.Os
.Sh NAME
.Nm wsfontload
.Nd load or delete a font bitmap from a wscons
display device
.Sh SYNOPSIS
.Nm wsfontload
.Bk -words
.Op Fl Bbl
.Op Fl e Ar encoding
.Op Fl f Ar file
.Op Fl h Ar height
.Op Fl N Ar name
.Op Fl w Ar width
.Op Ar fontfile
.Ek
.Nm wsfontload
.Op Fl f Ar file
.Fl d Ar slot
.Sh DESCRIPTION
The
.Nm
utility loads a font bitmap to a wscons device if the device driver
supports it.
The font gets assigned a name in this process which it can be referred to
by later for use on a display screen.
The font is loaded from the specified
.Ar fontfile ,
or from
.Pa stdin
if
.Ar fontfile
is not provided.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl B
Specifies that the font data is ordered right-to-left byte wise.
The default is left-to-right.
.It Fl b
Specifies that the font data is ordered right-to-left bit wise.
The default is left-to-right.
.It Fl d Ar slot
Specifies to delete the font from the given slot number, no other
arguments should be specified.  Will fail if font deletion is not
supported by the device, or if the font is in use.
.It Fl e Ar encoding
Sets the encoding of the font.
This can be either a symbolic abbreviation or a numeric value.
Currently recognized abbreviations are
.Dq iso
for ISO-8859-1 encoding,
.Dq ibm
for IBM encoded fonts, and
.Dq pcvt
for the custom encoding of the supplemental fonts which came with the BSD
PCVT console driver.
Per default,
.Dq iso
is assumed.
.It Fl f Ar file
Specify the control device of the wscons display to operate on.
Default is
.Pa /dev/ttyCcfg .
.It Fl h Ar height
Sets the height of a font character in pixels.
Default is 16.
.It Fl l
Specifies to print out a list of loaded fonts, no other
arguments should be specified.
.It Fl N Ar name
Specifies a name which can be used later to refer to the font.
If none is given, the
.Ar fontfile
name is used to create one.
.It Fl w Ar width
Sets the width of a font character in pixels.
Default is 8.
.El
.Pp
No font files are provided with the wscons framework.
The fonts installed by PCVT can be used instead, as can raw font files from
other operating system distributions.
.Sh FILES
.Bl -tag -width "/usr/share/misc/pcvtfonts/XX" -compact
.\" .It Pa /etc/wscons.conf
.\" wscons configuration file
.It Pa /usr/share/misc/pcvtfonts/
fonts directory.
.El
.Sh EXAMPLES
.Bd -literal
# wsfontload -N myname -h 8 -e ibm \e
	/usr/share/misc/pcvtfonts/vt220l.808
.Ed
.Pp
Load the IBM-encoded 8x8-font from the PCVT distribution.
This (or another 8x8-font) is necessary to use the 50-line screen type on
.Xr vga 4
displays.
.Pp
.Li "# wsfontload -d 3"
.Pp
Delete the font in currently loaded in slot 3.
.Sh SEE ALSO
.Xr wscons 4 ,
.Xr wsconscfg 8 ,
.Xr wsconsctl 8 ,
.Xr wssetfont 8
.Sh HISTORY
The
.Nm
program appeared in
.Ox 2.8 .
.Sh BUGS
Many features are missing.


src/usr.sbin/wsfontload/wsfontload.c
/* $MirOS: src/usr.sbin/wsfontload/wsfontload.c,v 1.3 2005/12/04 15:02:34 tg Exp $ */
/* $OpenBSD: wsfontload.c,v 1.10 2005/05/27 05:03:47 millert Exp $ */
/* $NetBSD: wsfontload.c,v 1.2 2000/01/05 18:46:43 ad Exp $ */

/*
 * Copyright (c) 1999
 *	Matthias Drochner.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed for the NetBSD Project
 *	by Matthias Drochner.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/ioctl.h>

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <err.h>

#include <dev/wscons/wsconsio.h>

__RCSID("$MirOS: src/usr.sbin/wsfontload/wsfontload.c,v 1.3 2005/12/04 15:02:34 tg Exp $");

#define DEFDEV		"/dev/ttyCcfg"
#define DEFWIDTH	8
#define DEFHEIGHT	16
#define DEFENC		WSDISPLAY_FONTENC_ISO
#define DEFBITORDER	WSDISPLAY_FONTORDER_L2R
#define DEFBYTEORDER	WSDISPLAY_FONTORDER_L2R

int main(int, char**);
static void usage(void);
static int getencoding(char *);

static void
usage(void)
{
	extern char *__progname;

	(void)fprintf(stderr,
	    "usage: %s [-Bbl] [-e encoding] [-f file] [-h height] [-N name]\n"
	    "       %*s [-w width] [fontfile]\n"
	    "       %s [-f file] -d slot\n",
	    __progname, (int)strlen(__progname), "", __progname);
	exit(1);
}

static const
struct {
	const char *name;
	int val;
} encodings[] = {
	{"iso",  WSDISPLAY_FONTENC_ISO},
	{"ibm",  WSDISPLAY_FONTENC_IBM},
	{"pcvt", WSDISPLAY_FONTENC_PCVT},
	{"iso7", WSDISPLAY_FONTENC_ISO7},
#ifdef WSDISPLAY_FONTENC_SONY
	{"sony", WSDISPLAY_FONTENC_SONY},
#endif
};

int
main(int argc, char *argv[])
{
	char *wsdev, *p;
	struct wsdisplay_font f;
	int c, res, wsfd, ffd, list, del, slot, i;
	size_t len;
	void *buf;

	wsdev = DEFDEV;
	f.index = -1;
	f.fontwidth = DEFWIDTH;
	f.fontheight = DEFHEIGHT;
	f.firstchar = 0;
	f.numchars = 256;
	f.stride = 0;
	f.encoding = DEFENC;
	f.name[0] = 0;
	f.bitorder = DEFBITORDER;
	f.byteorder = DEFBYTEORDER;

	list = 0; del = 0;
	while ((c = getopt(argc, argv, "bB:d:e:f:h:lN:w:")) != -1) {
		switch (c) {
		case 'f':
			wsdev = optarg;
			break;
		case 'w':
			if (sscanf(optarg, "%d", &f.fontwidth) != 1)
				errx(1, "invalid font width of %d",
				    f.fontwidth);
			break;
		case 'h':
			if (sscanf(optarg, "%d", &f.fontheight) != 1)
				errx(1, "invalid font height of %d",
				    f.fontheight);
			break;
		case 'e':
			f.encoding = getencoding(optarg);
			break;
		case 'l':
			list++;
			break;
		case 'd':
		        del++;
			if (sscanf(optarg, "%d", &slot) != 1)
			  errx(1, "invalid font number %d", slot);
			break;
		case 'N':
			strlcpy(f.name, optarg, WSFONT_NAME_SIZE);
			break;
		case 'b':
			f.bitorder = WSDISPLAY_FONTORDER_R2L;
			break;
		case 'B':
			f.byteorder = WSDISPLAY_FONTORDER_R2L;
			break;
		case '?':
		default:
			usage();
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if ((list + del) > 1)
		usage();
	if ((list || del) && argc)
		usage();

	if (argc > 1)
		usage();

	wsfd = open(wsdev, O_RDWR, 0);
	if (wsfd < 0)
		err(2, "open %s", wsdev);

	if (list) {
		i = 0;
		p = " # Name             Encoding  W  H";
		do {
			f.index = i;
			res = ioctl(wsfd, WSDISPLAYIO_LSFONT, &f);
			if (res == 0) {
				if (f.name[0]) {
					if (p) {
						puts(p);
						p = NULL;
					}
					printf("%2d %-16s %8s %2d %2d\n",
					    f.index, f.name,
					    encodings[f.encoding].name,
					    f.fontwidth, f.fontheight);
				}
			}
			i++;
		} while(res == 0);

		return (0);
	}

	if (del) {
		f.index = slot;
		res = ioctl(wsfd, WSDISPLAYIO_LSFONT, &f);
			printf("deleting font %s from slot %d\n",
			       f.name, slot);
			res = ioctl(wsfd, WSDISPLAYIO_DELFONT, &f);
			if (res)
				printf("unable to delete font\n");
			return res;
	}

	if (argc > 0) {
		ffd = open(argv[0], O_RDONLY, 0);
		if (ffd < 0)
			err(4, "open %s", argv[0]);
		if (!*f.name)
			strlcpy(f.name, argv[0], WSFONT_NAME_SIZE);
	} else
		ffd = 0;

	if (!f.stride)
		f.stride = (f.fontwidth + 7) / 8;
	len = f.fontheight * f.numchars * f.stride;
	if (!len)
		errx(1, "invalid font size");

	buf = malloc(len);
	if (!buf)
		errx(1, "malloc");
	res = read(ffd, buf, len);
	if (res < 0)
		err(4, "read %s", argv[0]);
	if (res != len)
		errx(4, "short read on %s", argv[0]);

	f.data = buf;

	res = ioctl(wsfd, WSDISPLAYIO_LDFONT, &f);
	if (res < 0)
		err(3, "WSDISPLAYIO_LDFONT");

	return (0);
}

static int
getencoding(char *name)
{
	int i;

	for (i = 0; i < sizeof(encodings) / sizeof(encodings[0]); i++)
		if (!strcmp(name, encodings[i].name))
			return (encodings[i].val);

	if (sscanf(name, "%d", &i) != 1)
		errx(1, "invalid encoding");
	return (i);
}
